exports: Add a separate bucket for realm exports.

prakhar1144 opened the pull request:
This PR extracts the first 6 commits of #21725 

While rebasing, I made few changes which were impacted by these commits:
* https://github.com/zulip/zulip/commit/214bd4ed88f6fb245dfa018fc00cbee5258e8169
* https://github.com/zulip/zulip/commit/3aea67a8ede6b8493f1faaa361556e376e94e203

I have highlighted the changes I made as comments below.

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

prakhar1144 commented on zerver/lib/upload/s3.py:
```diff
@@ -108,16 +109,28 @@ def upload_content_to_s3(
+    """
+    global BOTO_CLIENT
+    if BOTO_CLIENT is None:
+        BOTO_CLIENT = get_bucket(settings.S3_AUTH_UPLOADS_BUCKET).meta.client
```
The way we get `boto_client` is different as compared to #21725 because of the changes we made in https://github.com/zulip/zulip/commit/3aea67a8ede6b8493f1faaa361556e376e94e203

prakhar1144 commented on zerver/lib/upload/s3.py:
```diff
@@ -417,16 +421,44 @@ def get_export_tarball_url(self, realm: Realm, export_path: str) -> str:
+                },
+                # Expires in one week, the longest allowed by AWS
+                ExpiresIn=60 * 60 * 24 * 7,
+            )
```
Is this intended to not do the same 
```
            # Strip off the signing query parameters, since this URL is public
            return urlsplit(signed_url)._replace(query="").geturl()
```
operation here as we do in the `else` block below ?

alexmv replied:
Correct.  These URLs are actually time-bounded, which means that we can start expiring the uploads.  The exports which are shoved into the avatars bucket are protected only by security-through-obscurity.

alexmv commented on zerver/lib/upload/s3.py:
```diff
@@ -402,12 +402,13 @@ def upload_single_emoji_image(
     def get_export_tarball_url(self, realm: Realm, export_path: str) -> str:
         client = self.avatar_bucket.meta.client
+        if export_path.startswith("/"):
+            export_path = export_path.removeprefix("/")
```
The `startswith` check is unnecessary here.

timabbott replied:
Yeah, `removeprefix` removes the prefix if and only if it is present.

prakhar1144 replied:
Fixed it.

alexmv commented on zerver/lib/upload/s3.py:
```diff
@@ -417,16 +421,44 @@ def get_export_tarball_url(self, realm: Realm, export_path: str) -> str:
-        # Strip off the signing query parameters, since this URL is public
-        return urlsplit(signed_url)._replace(query="").geturl()
+        if self.export_bucket:
+            if export_path.startswith("exports/"):
+                # Fix old data if the row was created when an export
+                # bucket was not in use:
+                export_path = export_path[len("exports/") :]
```
```suggestion
            # Fix old data if the row was created when an export bucket was not in use
            export_path = export_path.removeprefix("exports/")
```

prakhar1144 commented:
Updates in the recent push:

* Addressed the review comments -- see respective comments.
* Added two new tests -- `test_tarball_upload_avatar_bucket_download_export_bucket` & `test_tarball_upload_export_bucket_download_avatar_bucket` -- See last commit (test_upload_s3)

timabbott commented on zerver/lib/upload/s3.py:
```diff
@@ -145,6 +145,10 @@ class S3UploadBackend(ZulipUploadBackend):
         self.uploads_bucket = get_bucket(settings.S3_AUTH_UPLOADS_BUCKET)
+        self.export_bucket: Bucket | None = None
+        if settings.S3_EXPORT_BUCKET:
+            self.export_bucket = get_bucket(settings.S3_EXPORT_BUCKET)
```
Would it work to just have `export_bucket` equal `uploads_bucket` rather than `None` if `S3_EXPORT_BUCKET` is not set? I feel like that might avoid a big conditional duplicating code below, but maybe I'm missing something.

alexmv replied:
I guess?  I liked the cleanliness of the path without `exports` in it twice.

timabbott replied:
OK. We can always refactor the internals of this later if it seems warranted.

timabbott commented:
Looks good to me aside from the question above.

timabbott commented:
Merged, thanks @prakhar1144 and @alexmv!
